pkg/proxmox/instance_addresses.go

```diff
@@ -0,0 +1,303 @@
+	providedAddresses := []string{}
+	if providedIP != "" {
+		providedAddresses = strings.Split(providedIP, ",")
+	}
+
+	addresses := []v1.NodeAddress{
+		{Type: v1.NodeHostName, Address: node.Name},
+	}
+
+	for _, address := range providedAddresses {
+		if address = strings.TrimSpace(address); address != "" {
+			parsedAddress := net.ParseIP(address)
+			if parsedAddress != nil {
+				addresses = append(addresses, v1.NodeAddress{
+					Type:    v1.NodeInternalIP,
+					Address: parsedAddress.String(),
+				})
+			} else {
+				klog.Warningf("Ignoring invalid provided address '%s' for node %s", address, node.Name)
+			}
+		}
+	}
+
+	if i.networkOpts.Mode == NetworkModeDefault {
+		// If the network mode is 'default', we only return the provided IPs.
+		klog.V(4).InfoS("instances.addresses() returning provided IPs", "node", klog.KObj(node))
+
+		return addresses
+	}
+
+	if i.networkOpts.Mode == NetworkModeOnlyHost || i.networkOpts.Mode == NetworkModeAuto {
+		newAddresses, err := i.retrieveHostAddresses(ctx)
+		if err != nil {
+			klog.ErrorS(err, "Failed to retrieve host addresses")
+		} else {
+			addToNodeAddresses(&addresses, newAddresses...)
+		}
+	}
+
+	if i.networkOpts.Mode == NetworkModeOnlyQemu || i.networkOpts.Mode == NetworkModeAuto {
+		newAddresses, err := i.retrieveQemuAddresses(ctx, vmRef, region)
+		if err != nil {
+			klog.ErrorS(err, "Failed to retrieve host addresses")
+		} else {
+			addToNodeAddresses(&addresses, newAddresses...)
+		}
+	}
+
+	// Remove addresses that match the ignored CIDRs
+	if len(i.networkOpts.IgnoredCIDRs) > 0 {
+		var removeableAddresses []v1.NodeAddress
+
+		for _, addr := range addresses {
+			ip := net.ParseIP(addr.Address)
+			if ip != nil && isAddressInCIDRList(i.networkOpts.IgnoredCIDRs, ip) {
+				removeableAddresses = append(removeableAddresses, addr)
+			}
+		}
+
+		removeFromNodeAddresses(&addresses, removeableAddresses...)
+	}
+
+	sortNodeAddresses(addresses, i.networkOpts.SortOrder)
+
+	return addresses
+}
+
+// retrieveQemuAddresses retrieves the addresses from the QEMU agent
+func (i *instances) retrieveQemuAddresses(ctx context.Context, vmRef *pxapi.VmRef, region string) ([]v1.NodeAddress, error) {
+	var addresses []v1.NodeAddress
+
+	klog.V(4).InfoS("retrieveQemuAddresses() retrieving addresses from QEMU agent")
+
+	r, err := i.getInstanceNics(ctx, vmRef, region)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, nic := range r {
+		for _, ip := range nic.IpAddresses {
+			i.processIP(ctx, &addresses, ip)
+		}
+	}
+
+	klog.V(4).InfoS("retrieveQemuAddresses() retrieved instance nics", "nics", r)
+
+	return addresses, nil
+}
+
+// retrieveHostAddresses retrieves the host addresses from the host network's network interfaces
+func (i *instances) retrieveHostAddresses(ctx context.Context) ([]v1.NodeAddress, error) {
+	klog.V(4).InfoS("retrieveHostAddresses() returning node status addresses")
+
+	var addresses []v1.NodeAddress
+
+	interfaces, err := net.Interfaces()
```

Comment:
As I understand it, `net.Interfaces()` returns the local IPs of the machine where the go-code is running. Therefore, it can't reliably provide the actual IPs of the external nodes.

Reply:
Agreed, but I think that should be something the admins should worry about or know. Would an acceptable compromise be adding doco about it?

Reply:
to use net.Interfaces() we need daemonset on all nodes, better to not have this option.

Reply:
Removed in latest